### ISSUE TITLE
CopyFiles respects the casing of the constructor argument, not the file system

### ISSUE THREAD

**Issue by deanebarker** [statiqdev/Statiq#202]
Found some weirdness with `CopyFiles`.

It doesn't copy with the same casing as the source folder name, _it copies with the same casing as the constructor_.

Consider a folder in the input directory called "assets" (lower-case).

`CopyFiles("Assets/*")`

That will copy it to the output folder as "Assets" (upper-case).  When I lower-cased the call to the `CopyFiles` constructor, _then_ it copied as lower-case.

The built-in preview server is not case-sensitive.  Amazon S3 is.  When I synced with S3, stuff was broken.

I don't know what expected behavior should be here. My gut says we should copy the casing of the source folder, but I was clearly inconsistent because my call to `CopyFiles` didn't match the casing of the file system.  But should that matter?  The Windows file system is case-insensitive.

I don't know that any one answer is any righter than another, but I wonder if there's a way we can document the behavior, and maybe have some constructor switches to be more explicit?

**Comment by daveaglick** [comment]
Two thoughts:
- I feel pretty strongly the copy should match the case of the original, _unless_ it's going into a folder that already exists _and_ the file system is case insensitive. If the destination path doesn't exist or the file system is case sensitive, any generated path components (including the file name) should match the source. In that case, I'd consider this a bug.
- This was a great catch, and exactly the kind of thing the new IO abstraction is designed to help. The new file system class has a settable property to indicate of the file system is case sensitive or not. In addition, once the modules get converted over to use it some of this will be taken care of in the background (the infrastructure has been partly converted, but modules are still using string-based paths right now). It'll become even more important when globbing is introduced for more complex search patterns.

TL;DR - this should be a bug and I'll keep it open until after IO rewrite is complete to verify.

**Comment by daveaglick** [comment]
So I've had to backtrack on this one a little bit. Consider the following:
- The new file provider classes related to #214 can pull files from various places like the local file system, a network share, a GitHub repository, .zip file, etc. Each of these sources may have different casing rules.
- The way multiple input paths _from multiple providers_ can be combined into a virtual file system that aggregates all of them means that we may be mixing directories with different case sensitivity into a single virtual directory.

Now suppose we're trying to evaluate a globbing pattern against that virtual directory that includes underlying providers which are both case sensitive and case insensitive. How would the globbing engine process matches? It would be really confusing... So (at least for now) all providers and associated file operations are case sensitive, even if the underlying file system is not. This lets us approach the problem with a "least common denominator" approach.

Of course, we may need to refine this strategy in practice as we start building sites once it's in place.

**Comment by daveaglick** [comment]
Now that the new I/O stuff is done, I can address this bug. It turns out that _all_ paths in Wyam are now case sensitive (see http://wyam.io/getting-started/io#case-sensitivity). This means that in the example above, if the `CopyFiles` search path hadn't matched the case of the folder on disk, the files would never have been copied. Note that the underlying file system may not be case sensitive, in which case it would still copy. In this situation however, it should always copy with the original case.
